src/intermediate.rs

```diff
@@ -79,20 +95,130 @@ impl<'a> IntermediateTypes<'a> {
             || self.generic_instances.contains_key(ident)
     }
 
-    pub fn type_aliases(&self) -> &BTreeMap<AliasIdent, (RustType, bool, bool)> {
+    pub fn type_aliases(&self) -> &BTreeMap<AliasIdent, AliasInfo> {
         &self.type_aliases
     }
 
     pub fn rust_structs(&self) -> &BTreeMap<RustIdent, RustStruct> {
         &self.rust_structs
     }
 
-    fn aliases() -> BTreeMap<idents::AliasIdent, (RustType, bool, bool)> {
+    /// For each scope, which other scopes are referenced, and which structs are referenced
+    pub fn scope_references(
+        &self,
+        wasm: bool,
+    ) -> BTreeMap<String, BTreeMap<String, BTreeSet<RustIdent>>> {
+        // we only want to mark TOP-LEVEL references without recursing into those types
+        // which is why we don't use visit_types() here
+        let mut refs = BTreeMap::new();
+        fn set_ref(
+            refs: &mut BTreeMap<String, BTreeMap<String, BTreeSet<RustIdent>>>,
+            types: &IntermediateTypes,
+            current_scope: &str,
+            rust_ident: &RustIdent,
+        ) {
+            let ref_scope = types.scope(rust_ident);
+            if current_scope != ref_scope {
+                refs.entry(current_scope.to_owned())
+                    .or_default()
+                    .entry(ref_scope.to_owned())
+                    .or_default()
+                    .insert(rust_ident.clone());
+            }
+        }
+        fn mark_refs(
+            refs: &mut BTreeMap<String, BTreeMap<String, BTreeSet<RustIdent>>>,
+            types: &IntermediateTypes,
+            wasm: bool,
+            current_scope: &str,
+            ty: &RustType,
+        ) {
+            match &ty.conceptual_type {
+                ConceptualRustType::Alias(alias_ident, _alias_ty) => {
+                    if let AliasIdent::Rust(rust_ident) = alias_ident {
+                        set_ref(refs, types, current_scope, rust_ident);
+                    }
+                }
+                ConceptualRustType::Rust(rust_ident) => {
+                    set_ref(refs, types, current_scope, rust_ident)
+                }
+                ConceptualRustType::Array(elem_ty) => {
+                    if wasm && !elem_ty.directly_wasm_exposable() {
+                        // TODO: we should be doing array wrappers where they are declared or used
+                        // but for the latter, what to do if multiple places use it? default to lib?
```

Comment:
For the TODO here and below, it may be good to create a Github issue for this and reference it in the code

Reply:
Will fix in next PR as this PR already contains a lot of different stuff